### PR TITLE
[fix] (routine load)Routine load task doesn't reallocate when previous BE is down.

### LINKED ISSUE #8823: [Bug] [Routine Load] routine load won't consume part of KAFKA partitions when a BE down
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Version

master 

### What's Wrong?

An routine load doesn't consume part of KAFKA partitions when a BE down

### What You Expected?

consume well.

### How to Reproduce?

start an routine load, and make a BE down, but don't drop it. (the routine load task should be assigned to this BE)

### Anything Else?

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by Henry2SS**
# Proposed changes

Issue Number: close #8823 

## Problem Summary:

if previous be is not alive, should assigned another available BE instead.

## Checklist(Required)

1. Does it affect the original behavior: (No)
2. Has unit tests been added: (No Need)
3. Has document been added or modified: (No Need)
4. Does it need to update dependencies: (No)
5. Are there any changes that cannot be rolled back: (No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: docs/zh-CN/administrator-guide/config/fe_config.md [hunk 1]
```diff
@@ -2220,3 +2220,13 @@ load 标签清理器将每隔 `label_clean_interval_second` 运行一次以清
 是否可以动态配置：false
 
 是否为 Master FE 节点独有的配置项：false
+
+### routine_load_task_reallocate_times
+
+一个routine load task尝试分配的最大次数，达到这个值，任务会重新分配
```
Comment (caiconghui) [inline docs/zh-CN/administrator-guide/config/fe_config.md:2226]:
```suggestion
一个routine load task尝试分配到同一个be的最大失败重试次数，达到这个值，任务会重新分配
```

File: docs/en/administrator-guide/config/fe_config.md [hunk 2]
```diff
@@ -2197,3 +2197,13 @@ Default: 10
 Is it possible to dynamically configure: false
 
 Is it a configuration item unique to the Master FE node: false
+
+### routine_load_task_reallocate_times
+
+The maximum times of a routine load do allocate before re-allocate to another BE. Once reached this number, the task will be re-allocate to another BE.
```
Comment (caiconghui) [inline docs/en/administrator-guide/config/fe_config.md:2203]: same as Chinese document

File: docs/zh-CN/administrator-guide/config/fe_config.md [hunk 3]
```diff
@@ -2220,3 +2220,13 @@ load 标签清理器将每隔 `label_clean_interval_second` 运行一次以清
 是否可以动态配置：false
 
 是否为 Master FE 节点独有的配置项：false
+
+### routine_load_task_reallocate_times
```
Comment (caiconghui) [inline docs/zh-CN/administrator-guide/config/fe_config.md:2224]:
```suggestion
### max_retry_reallocate_times_for_same_be
```

File: fe/fe-core/src/main/java/org/apache/doris/load/routineload/RoutineLoadTaskScheduler.java [hunk 4]
```diff
@@ -289,6 +289,13 @@ private void submitTask(long beId, TRoutineLoadTask tTask) throws LoadException
     private boolean allocateTaskToBe(RoutineLoadTaskInfo routineLoadTaskInfo) throws LoadException {
         long beId = routineLoadManager.getAvailableBeForTask(routineLoadTaskInfo.getPreviousBeId(), routineLoadTaskInfo.getClusterName());
         if (beId == -1L) {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/load/routineload/RoutineLoadTaskScheduler.java:291]: I think we can just modify the `getAvailableBeForTask()` method?
In that method: check if previous Backend is alive, if yes, go on. If no, find a new one.

Comment (morningman) [reply fe/fe-core/src/main/java/org/apache/doris/load/routineload/RoutineLoadTaskScheduler.java:291]: So that we don't need to add config for this.

Comment (Henry2SS) [reply fe/fe-core/src/main/java/org/apache/doris/load/routineload/RoutineLoadTaskScheduler.java]: Done.

File: fe/fe-core/src/main/java/org/apache/doris/load/routineload/RoutineLoadManager.java [hunk 5]
```diff
@@ -410,36 +411,43 @@ public long getMinTaskBeId(String clusterName) throws LoadException {
...
     // throw exception if unrecoverable errors happen.
-    public long getAvailableBeForTask(long previoudBeId, String clusterName) throws LoadException {
+    public long getAvailableBeForTask(long previousBeId, String clusterName) throws LoadException {
         List<Long> beIdsInCluster = Catalog.getCurrentSystemInfo().getClusterBackendIds(clusterName, true);
         if (beIdsInCluster == null) {
             throw new LoadException("The " + clusterName + " has been deleted");
         }
 
-        if (previoudBeId != -1L && !beIdsInCluster.contains(previoudBeId)) {
+        if (previousBeId != -1L && !beIdsInCluster.contains(previousBeId)) {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/load/routineload/RoutineLoadManager.java:420]: If previous BE is down, this method still return -1 and task will be failed to allocated.
So nothing change?

Comment (Henry2SS) [reply fe/fe-core/src/main/java/org/apache/doris/load/routineload/RoutineLoadManager.java]: Fixed.

File: fe/fe-core/src/main/java/org/apache/doris/load/routineload/RoutineLoadManager.java [hunk 6]
```diff
@@ -410,36 +411,39 @@ public long getMinTaskBeId(String clusterName) throws LoadException {
...
-                } else if (beIdToConcurrentTasks.containsKey(previoudBeId)) {
-                    idleTaskNum = beIdToMaxConcurrentTasks.get(previoudBeId) - beIdToConcurrentTasks.get(previoudBeId);
-                } else {
-                    idleTaskNum = Config.max_routine_load_task_num_per_be;
-                }
-                if (idleTaskNum > 0) {
-                    return previoudBeId;
+            if (previousBeId != -1L) {
+                // get the previousBackend info
+                Backend previousBackend = Catalog.getCurrentSystemInfo().getBackend(previousBeId);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/load/routineload/RoutineLoadManager.java:428]: `previousBackend` maybe null

Comment (Henry2SS) [reply fe/fe-core/src/main/java/org/apache/doris/load/routineload/RoutineLoadManager.java:428]: Fixed.

**Review by caiconghui — DISMISSED** [review]
LGTM

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**